core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/BestSolutionHolder.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.solver;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.change.ProblemChange;
+
+final class BestSolutionHolder<Solution_> {
+
+    private final Lock problemChangesLock = new ReentrantLock();
+    private final AtomicReference<VersionedBestSolution> versionedBestSolutionRef = new AtomicReference<>();
+    private final SortedMap<BigInteger, List<CompletableFuture<Void>>> problemChangesPerVersion =
+            new TreeMap<>();
+    private BigInteger currentVersion = BigInteger.ZERO;
+
+    boolean isEmpty() {
+        return versionedBestSolutionRef.get() == null;
+    }
+
+    /**
+     * @return the last best solution together with problem changes the solution contains.
+     */
+    BestSolutionContainingProblemChanges<Solution_> take() {
+        VersionedBestSolution versionedBestSolution = versionedBestSolutionRef.getAndSet(null);
+        if (versionedBestSolution == null) {
+            return null;
+        }
+        SortedMap<BigInteger, List<CompletableFuture<Void>>> containedProblemChangesPerVersion =
+                problemChangesPerVersion.headMap(versionedBestSolution.getVersion().add(BigInteger.ONE));
+
+        List<CompletableFuture<Void>> containedProblemChanges = containedProblemChangesPerVersion.values()
+                .stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        containedProblemChangesPerVersion.keySet().forEach(problemChangesPerVersion::remove);
```

Comment:
Why not just do it in one `forEach(...)`? I see the intention for separation of concerns, but this code is simple enough that I think the double iteration is not worth it.